drivers/input/input_paw32xx.c

```diff
@@ -0,0 +1,443 @@
+#define PAW32XX_SLEEP2		0x0b
+#define PAW32XX_SLEEP3		0x0c
+#define PAW32XX_CPI_X		0x0d
+#define PAW32XX_CPI_Y		0x0e
+#define PAW32XX_DELTA_XY_HI	0x12
+#define PAW32XX_MOUSE_OPTION	0x19
+
+#define PRODUCT_ID_PAW32XX 0x30
+#define SPI_WRITE BIT(7)
+
+#define MOTION_STATUS_MOTION BIT(7)
+#define OPERATION_MODE_SLP_ENH	BIT(4)
+#define OPERATION_MODE_SLP2_ENH	BIT(3)
+#define OPERATION_MODE_SLP_MASK (OPERATION_MODE_SLP_ENH | OPERATION_MODE_SLP2_ENH)
+#define CONFIGURATION_PD_ENH BIT(3)
+#define CONFIGURATION_RESET BIT(7)
+#define WRITE_PROTECT_ENABLE 0x00
+#define WRITE_PROTECT_DISABLE 0x5a
+#define MOUSE_OPTION_MOVX_INV_BIT 3
+#define MOUSE_OPTION_MOVY_INV_BIT 4
+
+#define PAW32XX_DATA_SIZE_BITS 12
+
+#define RESET_DELAY_MS 2
+
+#define RES_STEP 38
+#define RES_MIN (16 * RES_STEP)
+#define RES_MAX (127 * RES_STEP)
+
+struct paw32xx_config {
+	struct spi_dt_spec spi;
+	struct gpio_dt_spec motion_gpio;
+	uint16_t axis_x;
+	uint16_t axis_y;
+	int16_t res_cpi;
+	bool invert_x;
+	bool invert_y;
+	bool force_awake;
+};
+
+struct paw32xx_data {
+	const struct device *dev;
+	struct k_work motion_work;
+	struct gpio_callback motion_cb;
+};
+
+static int paw32xx_read(const struct device *dev,
+			uint8_t addr, uint8_t *value, uint8_t len)
+{
+	const struct paw32xx_config *cfg = dev->config;
+
+	const struct spi_buf tx_buf = {
+		.buf = &addr,
+		.len = sizeof(addr),
+	};
+	const struct spi_buf_set tx = {
+		.buffers = &tx_buf,
+		.count = 1,
+	};
+
+	struct spi_buf rx_buf[] = {
+		{
+			.buf = NULL,
+			.len = sizeof(addr),
+		},
+		{
+			.buf = value,
+			.len = len,
+		},
+	};
+	const struct spi_buf_set rx = {
+		.buffers = rx_buf,
+		.count = ARRAY_SIZE(rx_buf),
+	};
+
+	return spi_transceive_dt(&cfg->spi, &tx, &rx);
+}
+
+static int paw32xx_read_reg(const struct device *dev, uint8_t addr, uint8_t *value)
+{
+	return paw32xx_read(dev, addr, value, 1);
+}
+
+static int paw32xx_write_reg(const struct device *dev, uint8_t addr, uint8_t value)
+{
+	const struct paw32xx_config *cfg = dev->config;
+
+	uint8_t write_buf[] = {addr | SPI_WRITE, value};
+	const struct spi_buf tx_buf = {
+		.buf = write_buf,
+		.len = sizeof(write_buf),
+	};
+	const struct spi_buf_set tx = {
+		.buffers = &tx_buf,
+		.count = 1,
+	};
+
+	return spi_write_dt(&cfg->spi, &tx);
+}
+
+static int paw32xx_update_reg(const struct device *dev, uint8_t addr, uint8_t mask, uint8_t value)
+{
+	uint8_t val;
+	int ret;
+
+	ret = paw32xx_read_reg(dev, addr, &val);
+	if (ret < 0) {
+		return ret;
+	}
+
+	val = (val & ~mask) | (value & mask);
+
+	ret = paw32xx_write_reg(dev, addr, val);
+	if (ret < 0) {
+		return ret;
+	}
+
+	return 0;
+}
+
+static void paw32xx_motion_work_handler(struct k_work *work)
+{
+	struct paw32xx_data *data = CONTAINER_OF(
+			work, struct paw32xx_data, motion_work);
+	const struct device *dev = data->dev;
+	const struct paw32xx_config *cfg = dev->config;
+	uint8_t val;
+	int16_t x, y;
+	int ret;
+
+	ret = paw32xx_read_reg(dev, 2, &val);
+	if (ret < 0) {
+		return;
```

Comment:
Wondering if some added logging would be a good idea here - it makes the code noisier (you could move it into the helper methods)

Reply:
not much into it to be honest, the device is rather simple

I did forget to use the definition here though (that "2"), fixed